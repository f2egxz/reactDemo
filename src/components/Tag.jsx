//-*- mode: react -*-
//-*- coding: utf-8 -*-

/**
 * Tag.jsx
 *
 * -TODO 标签,默认样式
 * -TODO 各种背景色的标签
 * -TODO 可删除的标签
 *
 * @author fe.gxz
 */


import React          from 'react';
import { classnames } from 'util/view-extra';
import Styles          from 'style/Tag.css';


function MapColorToStyle(color){
  const colorRegExp =/^#([0-9a-fA-F]{6}|[0-9a-fA-F]{3})$/
  const colorRegExpRGB = /^[rR][gG][Bb][Aa]?[\(]([\s]*(2[0-4][0-9]|25[0-5]|[01]?[0-9][0-9]?),){2}[\s]*(2[0-4][0-9]|25[0-5]|[01]?[0-9][0-9]?),?[\s]*(0\.\d{1,2}|1|0)?[\)]{1}$/g;
  if(color==='default'){
    return classnames(Styles.tag,Styles.dft)
  }else if(colorRegExp.test(color)||colorRegExpRGB.test(color)||typeof(color)==='string'){
    return classnames(Styles.tag,Styles.colors)
  }else{
    return classnames(Styles.tag,Styles.dft)
  }
}


/**
 * Tag
 *
 * 标签组件
 *
 * @param {String} color
 * @param {Boolean} closable
 * @param children
 * @return {XML}
 * @constructor
 */
function Tag({
  color,
  closable,
  children,
  handleClick
}){
  const rgb = color==='default'? '#f8f8f8':color;
  const close = closable?(<i className={classnames(Styles.delete)} onClick={handleClick}>x</i>) :'';
  const Tags = (
    <span className={MapColorToStyle(color)} style={{backgroundColor:rgb}}>
      {children}{close}
    </span>
  );
  return Tags
}

Tag.propTypes = {
  color    : React.PropTypes.string,
  closable : React.PropTypes.bool,
  handleClick : React.PropTypes.func
}

Tag.defaultProps = {
    color: 'default'
}

export default Tag
