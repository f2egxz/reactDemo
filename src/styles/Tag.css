/*-*- mode: css -*-*/
/*-*- coding: utf-8 -*-*/

/**
 * Tag.css
 *
 * 标签
 *
 * @author fe.gxz
 */

.tag {
  display: inline-block;
  padding:2px 8px;
  border-radius: 4px;
  cursor: default;
  color: #ffffff;
  background-color: red;
  margin: 4px;
}
.dft {
  border: 1px solid #000000;
  color: #333333;
}
.colors {
  padding:3px 9px;
}

.delete {
  position: relative;
  cursor: pointer;
  margin-left: 5px;
  font-size: 12px;
  top: -2px;
}

