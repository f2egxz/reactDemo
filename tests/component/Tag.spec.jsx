import React from 'react'
import Tag from 'component/Tag.jsx'
import render from 'react-test-renderer'

test('Tag test',()=>{
  const component = render.create(
    <Tag color="#108ee9" closable={true}>test</Tag>
  )
  let TagString = component.toJSON();
  expect(TagString).toMatchSnapshot();
})
