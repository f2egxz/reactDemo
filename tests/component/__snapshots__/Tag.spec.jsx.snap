exports[`test Tag test 1`] = `
<span
  className="tag colors"
  style={
    Object {
      "backgroundColor": "#108ee9",
    }
  }>
  test
  <i
    className="delete"
    onClick={undefined}>
    x
  </i>
</span>
`;
